src/metabase/models/recent_views.clj

```diff
@@ -0,0 +1,87 @@
+(ns metabase.models.recent-views
+  "The Recent Views table is used to track the most recent views of objects such as Cards, Tables and Dashboards for
+  each user."
+  (:require
+   [java-time :as t]
+   [metabase.util :as u]
+   [metabase.util.honey-sql-2 :as h2x]
+   [metabase.util.malli :as mu]
+   [metabase.util.malli.schema :as ms]
+   [methodical.core :as m]
+   [toucan2.core :as t2]))
+
+(doto :model/RecentViews
+  (derive :metabase/model))
+
+(m/defmethod t2/table-name :model/RecentViews
+  [_model]
+  :recent_views)
+
+(t2/define-before-insert :model/RecentViews
+  [log-entry]
+  (let [defaults {:timestamp :%now}]
+    (merge defaults log-entry)))
+
+(def ^:private ^:dynamic *recent-views-stored-per-user*
+  "The number of recently viewed items to keep per user. This should be larger than the number of items returned by the
+  /api/activity/recent_views endpoint, but it should still be lightweight to read all of a user's recent views at once."
+  100)
+
+(defn- view-ids-to-prune
+  "Returns a set of view IDs to prune from the RecentViews table so we only keep the most recent n views per user.
+  Ensures that we keep the most recent dashboard view for the user."
+  [prior-views n]
+  (if (< (count prior-views) n)
+    []
+    (let [ids-to-keep                    (map :id (take n prior-views))
+          ;; We want to make sure we keep the most recent dashboard view for the user
+          ids-to-prune                   (map :id (drop n prior-views))
+          most-recent-dashboard-id       (->> prior-views (filter #(= "dashboard" (:model %))) first :id)
+          pruning-most-recent-dashboard? ((set ids-to-prune) most-recent-dashboard-id)]
+      (if pruning-most-recent-dashboard?
+        (conj (remove #{most-recent-dashboard-id} (set ids-to-prune))
+              (last ids-to-keep))
+        ids-to-prune))))
+
+(mu/defn update-users-recent-views!
+  "Updates the RecentViews table for a given user with a new view, and prunes old views."
+  [user-id  :- [:maybe ms/PositiveInt]
+   model    :- [:or
+                [:enum :model/Card :model/Table :model/Dashboard]
+                :string]
+   model-id :- ms/PositiveInt]
+  (when user-id
+    (t2/with-transaction [_conn]
+      (t2/insert! :model/RecentViews {:user_id  user-id
+                                      ;; Lower-case the model name, since that's what the FE expects
+                                      :model    (u/lower-case-en (name model))
```

Comment:
Not a big deal, but should this lowercasing happen at the API level? I think in other places we store model as Capitalized Strings, so it's a bit inconsistent I think.

Reply:
That's fair, we should probably lower case on read instead of write. The previous versions of activity/view_log also used lower-case model names but I've been using capitalized strings for the new table since that's how we refer to models in the code base